### PR TITLE
Refactor controller status logic

### PR DESCRIPTION
**Pull request by MUzairS15**
Signed-off-by: MUzairS15 <muzair.shaikh810@gmail.com>

**Description**

This PR refactor the logic of status of Meshery Components wrt this [doc](https://docs.google.com/document/d/1q-aayRqx3QKIk2soTaTTTH-jmHcVXHwNYFsYkFawaME/edit#heading=h.ngupcd4j1pfm).
**Notes for Reviewers**

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [ ] Yes, I signed my commits.

### CODE REVIEW

**Comment by MUzairS15** [comment]
`GetPublicEndpoint` doesn't seem to work correctly, as even if broker-service has IP as pending this function returns an endpoint.
@leecalcote @sudo-NithishKarthik Can you explain this behaviour??

**Comment by leecalcote** [comment]
@MUzairS15 what is the output of the function in that scenario?

**Comment by leecalcote** [comment]
@sudo-NithishKarthik please review.

**Comment by MUzairS15** [comment]
> @MUzairS15 what is the output of the function in that scenario?
Screenshot 2022-07-28 at 8 35 07 PM
The output is the value in `External` field, though svc is not exposed.

**Comment by MUzairS15** [comment]
I think er should query for broker-svc instead of custom resource

**Comment by leecalcote** [comment]
> I think we should query for broker-svc instead of custom resource

Yes, potentially both. You're right that the service is ultimately what we need working / connected.

**Comment by leecalcote** [comment]
Starting with the broker svc is the right thing, though.

File: models/controllers/meshery_broker.go [hunk 1]
```diff
@@ -100,5 +104,22 @@ func (mb *mesheryBroker) GetPublicEndpoint() (string, error) {
...
+		return getImageVersionOfContainer(statefulSet.Spec.Template, "nats"), nil
+	}
+	return mb.version, nil
+}
+
+func getImageVersionOfContainer(container v1.PodTemplateSpec, containerName string) string {
+	var version string
+	for _, container := range container.Spec.Containers {
+		if strings.Compare(container.Name, containerName) == 0 {
+			version = strings.Split(container.Image, ":")[1]
```
Comment (Revolyssup) [inline models/controllers/meshery_broker.go:121]: Although there will be a tag, but still avoid accessing an index before checking length of array

**Comment by MUzairS15** [comment]
@Revolyssup

**Comment by sudo-NithishKarthik** [comment]
> GetPublicEndpoint doesn't seem to work correctly, as even if broker-service has IP as pending this function returns an endpoint.
@leecalcote @sudo-NithishKarthik Can you explain this behavior??


This might be because we have fallbacks that give the internal endpoint when it could not find the external endpoint

**Comment by MUzairS15** [comment]
> > GetPublicEndpoint doesn't seem to work correctly, as even if broker-service has IP as pending this function returns an endpoint.
> > @leecalcote @sudo-NithishKarthik Can you explain this behavior??
> 
> This might be because we have fallbacks that give the internal endpoint when it could not find the external endpoint

+1

**Comment by sudo-NithishKarthik** [comment]
In that case, we can either modify the function to not give internal endpoint or to return that information

File: models/controllers/helpers.go [hunk 2]
```diff
@@ -13,6 +16,16 @@ import (
 	"github.com/spf13/viper"
 )
 
+const BrokerPingEndpoint = "8222/connz"
+
+type Connections struct {
```
Comment (Revolyssup) [inline models/controllers/helpers.go:21]: Where will this struct be used?

**Comment by Revolyssup** [comment]
@MUzairS15 Also the lint tests failing

File: models/controllers/helpers.go (hunk 2 above)
Comment (MUzairS15) [reply models/controllers/helpers.go:21]: https://github.com/meshery/meshkit/pull/197/files#diff-a127b332cfc1dd1adf96e28240276fe52634ce187468c511bf31d30d55321a0eR112

**Review by Revolyssup — APPROVED** [review]
Just fix the lint checks and we are good to go

**Pull request merged**